### PR TITLE
[SPARK-25299] Use the shuffle writer plugin for the SortShuffleWriter.

### PR DESCRIPTION
**Pull request by mccheah**

### CODE REVIEW

File: core/src/main/scala/org/apache/spark/util/collection/ExternalSorter.scala [hunk 1]
```diff
@@ -722,6 +721,123 @@ private[spark] class ExternalSorter[K, V, C](
     lengths
   }
 
+  private def writeEmptyPartition(mapOutputWriter: ShuffleMapOutputWriter): Unit = {
+    var partitionWriter: ShufflePartitionWriter = null
+    try {
+      partitionWriter = mapOutputWriter.getNextPartitionWriter
```
Comment (ifilonenko) [inline core/src/main/scala/org/apache/spark/util/collection/ExternalSorter.scala]: Would you need to do `partitonWriter.toStream()`, as in `UnsafeShuffleWriter`, to ensure that the `outputFileStream` is created and an empty file exists. It seems be expected by the `UnsafeShuffleWriterSuite`, idk if it is the same here

Comment (mccheah) [reply core/src/main/scala/org/apache/spark/util/collection/ExternalSorter.scala]: That shouldn't be necessary, the `writer.close()` should properly know what to do if a stream was never created.

File: core/src/main/scala/org/apache/spark/util/collection/ShufflePartitionPairsWriter.scala [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+    }
+    objOut.writeKey(key)
+    objOut.writeValue(value)
+    writeMetrics.incRecordsWritten(1)
+  }
+
+  private def open(): Unit = {
+    // The contract is that the partition writer is expected to close its own streams, but
+    // the compressor will only flush the stream when it is specifically closed, not just
+    // when it's flushed.
```
Comment (yifeih) [inline core/src/main/scala/org/apache/spark/util/collection/ShufflePartitionPairsWriter.scala]: nit: reword `not just when it's actually flushed` -> `not when it's just flushed`.

File: core/src/main/scala/org/apache/spark/util/collection/ExternalSorter.scala [hunk 3]
```diff
@@ -722,6 +721,123 @@ private[spark] class ExternalSorter[K, V, C](
...
+          lengths(id) = partitionWriter.getNumBytesWritten
+        }
+        nextPartitionId = id + 1
+      }
+    }
+
+    // The iterator may have stopped short of opening a writer for every partition. So fill in the
+    // remaining empty partitions.
+    for (emptyPartition <- nextPartitionId until numPartitions) {
+      writeEmptyPartition(mapOutputWriter)
```
[the remark this answers is not in the document]
Comment (mccheah) [reply core/src/main/scala/org/apache/spark/util/collection/ExternalSorter.scala]: It depends on the contract we want to present to other plugin writers. I.e. do we make a contract that we open a writer for strictly every partition, even empty ones? Or do we say we open for the first N partitions where N is the last non-empty partition? My take is that we should have the contract that we always open a writer for every partition, empty or not, from 0 through numPartitions - 1. But, again, this shows the limitation of presenting an API that doesn't include the partition identifier explicitly when getting partition writers.

File: core/src/main/scala/org/apache/spark/util/collection/ShufflePartitionPairsWriter.scala (hunk 2 above)
Comment (mccheah) [reply core/src/main/scala/org/apache/spark/util/collection/ShufflePartitionPairsWriter.scala]: Think this is done

File: core/src/main/scala/org/apache/spark/util/collection/ExternalSorter.scala (hunk 3 above)
[the remark this answers is not in the document]
Comment (yifeih) [reply core/src/main/scala/org/apache/spark/util/collection/ExternalSorter.scala]: Ohh i see hmm yea I'm ok keeping it like this then. It does show more consistency for plugin implementers

**Comment by yifeih** [comment]
looks good to me!

**Comment by mccheah** [comment]
Ok I'm going to merge this - we have some incoming suggestions for the shuffle writer API but we should get the implementations with the existing APIs merged first. Thanks everyone who reviewed!

**Pull request merged**